### PR TITLE
[Terrain] Clipmap smart update initial version

### PR DESCRIPTION
**Pull request by jiaweig-amzn**
1. Created a special base pass for clipmap updating.
2. Clipmap compute pass will now be invoked only if there is a region to be updated.
3. The compute pass will distribute the region evenly to each compute thread. It is still dispatching at static thread numbers. We'll need some tests for dynamic numbers.
5. Fixed mod center no initialization in ClipmapBounds and added another test case for update region numbers.

To do:
1. There are still some float rounding issues within clipmap updating and sampling. One is by the precision of the rcp of the clipmap scale. Another is the actual view position doesn't match the clipmap center. Will fix these in later PR.
2. Add the margin for each clipmap to allow lazy update. This is partially done in the ClipmapCenter class using m_clipmapUpdateMultiple.

### CODE REVIEW

File: Gems/Terrain/Assets/Shaders/Terrain/TerrainDetailClipmapGenerationPass.azsl
Comment (mbalfour-amzn) [inline Gems/Terrain/Assets/Shaders/Terrain/TerrainDetailClipmapGenerationPass.azsl:38]: Suggestion: Add a comment and named constant to help explain why there's 6 regions.

File: Gems/Terrain/Assets/Shaders/Terrain/TerrainSrg.azsli
Comment (mbalfour-amzn) [inline Gems/Terrain/Assets/Shaders/Terrain/TerrainSrg.azsli:17]: Could you do `= MacroClipmapStackSizeMax * 6u;`? Also, this would again benefit from something explaining that this is related to having 6 regions (I'm guessing), and why there are 6 regions.

Might even consider a name based on the regions, not the number of regions, like 'MacroClipmapRegionStackSizeMax'.

File: Gems/Terrain/Code/Source/TerrainRenderer/TerrainClipmapManager.cpp
Comment (mbalfour-amzn) [inline Gems/Terrain/Code/Source/TerrainRenderer/TerrainClipmapManager.cpp:308]: m_fristTimeUpdate -> m_firstTimeUpdate

File: Gems/Terrain/Code/Source/TerrainRenderer/TerrainClipmapManager.cpp
Comment (mbalfour-amzn) [inline Gems/Terrain/Code/Source/TerrainRenderer/TerrainClipmapManager.cpp:388]: I assume the TODO will get removed _very_ shortly in an upcoming PR?

File: Gems/Terrain/Code/Source/TerrainRenderer/TerrainClipmapManager.h
Comment (mbalfour-amzn) [inline Gems/Terrain/Code/Source/TerrainRenderer/TerrainClipmapManager.h:199]: difined -> defined

File: Gems/Terrain/Assets/Shaders/Terrain/TerrainSrg.azsli
Comment (invertednormal) [reply Gems/Terrain/Assets/Shaders/Terrain/TerrainSrg.azsli:17]: I think this is a symptom of trying to make a single compute shader that does all the updating for every clipmap in every possible region all at once. This could be simplified so each invocation of the compute shader only cares about a single contiguous region on the clipmap.

File: Gems/Terrain/Code/Source/TerrainRenderer/Passes/TerrainClipmapComputePass.cpp
Comment (invertednormal) [inline Gems/Terrain/Code/Source/TerrainRenderer/Passes/TerrainClipmapComputePass.cpp:119]: There are utility functions in RPIUtils that could remove the need for most of this code.

**Review by invertednormal — CHANGES_REQUESTED** [review]
So it seems like the way things are going is we have a single compute shader that knows how to update all the clipmaps, and update multiple regions per clipmap based on the worst case scenario (like every clipmap needing to update 6 regions because a camera movement happened to require updates everywhere) This requires the compute shader to do lots of math to figure out the coordinate spaces, and go through every possible case to see if something needs to be updated. IMO this is way too complicated.

The common case will be just a few regions need to be updated. Larger clipmaps will only need to be updated every few frames (or even seconds for macro materials) with typical camera movement because each texel represent a fair amount of world space.

I think the main problem is that a single compute shader is dispatched that needs to figure out everything it might have to do, instead of multiple dispatches which have specific jobs to do. In other words, if the camera moves, and say 8 regions need to be updated in 3 clipmaps, then it would be much simpler to do 8 dispatches that specifically say update texture X in this clipmap region and this world region, then doing one dispatch that has to figure that out on its own (which requires massive arrays of textures and region data to prepare for any possible scenario)

File: Gems/Terrain/Assets/Shaders/Terrain/TerrainSrg.azsli
Comment (jiaweig-amzn) [reply Gems/Terrain/Assets/Shaders/Terrain/TerrainSrg.azsli:17]: Unfortunately, the shader compiler is not smart enough to figure out those constant sizes. Both `static const uint size = MacroClipmapStackSizeMax * 6u; array[size];` or `array[MacroClipmapStackSizeMax * 6u];` will fail at unresolved array dimensions.

File: Gems/Terrain/Code/Source/TerrainRenderer/TerrainClipmapManager.cpp
Comment (jiaweig-amzn) [reply Gems/Terrain/Code/Source/TerrainRenderer/TerrainClipmapManager.cpp:388]: I need to run some performance tests. I'll remove this and create a ticket to track it.
And some more discussion is going on, we may change this part completely.

File: Gems/Terrain/Assets/Shaders/Terrain/ClipmapComputeHelpers.azsli
Comment (jiaweig-amzn) [reply Gems/Terrain/Assets/Shaders/Terrain/ClipmapComputeHelpers.azsli:104]: Yes. I'll need to pass more data to solve the float rounding error issue as well. I'll get this part simplified in the next PR.

File: Gems/Terrain/Assets/Shaders/Terrain/TerrainMacroClipmapGenerationPass.azsl
Comment (jiaweig-amzn) [reply Gems/Terrain/Assets/Shaders/Terrain/TerrainMacroClipmapGenerationPass.azsl:29]: Per our discussion, the way you pointed out was my first attempt. There are many limitations on dispatching the exact region and invoking a compute shader. Such as how to bind different SRG, compiling SRG multiple times in a frame, building multiple passes internally. None of them can be resolved easily without breaking the current pass system. So I fell into this solution to add burdens for the compute shader to figure out the coordinates.

We'll have some discussion with pass system and upgrade it in the next iteration if it can. I wouldn't worry about the performance from comparing the test map I was running.

File: Gems/Terrain/Code/Source/TerrainRenderer/Passes/TerrainClipmapComputePass.cpp
Comment (jiaweig-amzn) [reply Gems/Terrain/Code/Source/TerrainRenderer/Passes/TerrainClipmapComputePass.cpp:119]: I'll try to clean this a bit.

**Review by invertednormal — APPROVED** [review]
Approving this to unblock, we'll chat about ways to simplify for next PR.

**Review by mbalfour-amzn — APPROVED** [review]

File: Gems/Terrain/Assets/Shaders/Terrain/ClipmapComputeHelpers.azsli
Comment (antonmic) [inline Gems/Terrain/Assets/Shaders/Terrain/ClipmapComputeHelpers.azsli:114]: You have this calculation:
`maxRenderRadius * 2.0 * clipmapScaleInv / TerrainSrg::m_clipmapData.m_clipmapSizeFloat`
repeated 3 times in this file, would be better to wrap it in a function with a descriptive name and maybe a comment or two about the details of the calculation.

File: Gems/Terrain/Assets/Shaders/Terrain/TerrainMacroClipmapGenerationPass.azsl
Comment (antonmic) [inline Gems/Terrain/Assets/Shaders/Terrain/TerrainMacroClipmapGenerationPass.azsl:56]: Recomputing the same value here just in X and Y, no? Can save/optimize out the duplicate calculation.

File: Gems/Terrain/Code/Source/TerrainRenderer/Passes/TerrainClipmapComputePass.cpp
Comment (antonmic) [inline Gems/Terrain/Code/Source/TerrainRenderer/Passes/TerrainClipmapComputePass.cpp:59]: You shouldn't be skipping the command list submission at this point if X or Y = 0. Instead you can override the pass's IsEnabled() function and check in there and return false if X or Y = 0. This way you will bypass the pass completely and not declare dependencies to the FrameGraph, which might cause errors if you do and then don't submit any dispatch item.

File: Gems/Terrain/Code/Source/TerrainRenderer/Passes/TerrainClipmapComputePass.cpp
Comment (antonmic) [inline Gems/Terrain/Code/Source/TerrainRenderer/Passes/TerrainClipmapComputePass.cpp:189]: This can also be moved to an IsEnabled() override so you can get the number of threads and return false if you don't need to run the pass.

File: Gems/Terrain/Code/Source/TerrainRenderer/Passes/TerrainClipmapComputePass.h
Comment (antonmic) [inline Gems/Terrain/Code/Source/TerrainRenderer/Passes/TerrainClipmapComputePass.h:45]: Why inherit from RenderPass and not ComputePass? This wasn't clear for me looking at the code.

File: Gems/Terrain/Code/Source/TerrainRenderer/Passes/TerrainClipmapComputePass.h
Comment (antonmic) [inline Gems/Terrain/Code/Source/TerrainRenderer/Passes/TerrainClipmapComputePass.h:44]: Class comment please - a bit about what this pass does so readers new to the system have some context.

**Review by antonmic — CHANGES_REQUESTED** [review]
Solid work overall, but should do the pass early out in the IsEnabled() function (which you can override) rather than in the build command lists function, since by that point you've already declared the scope and the attachment usages to the FrameGraph. Also not immediately clear why you inherit from RenderPass rather than ComputePass, it's not a blocker but would be nice if we could extend ComputePass to support the missing use case you have.

File: Gems/Terrain/Code/Source/TerrainRenderer/Passes/TerrainClipmapComputePass.h
Comment (jiaweig-amzn) [reply Gems/Terrain/Code/Source/TerrainRenderer/Passes/TerrainClipmapComputePass.h:45]: As some previous discussion, our first implementation was to dispatch multiple times. So I need some special operation in the ComputePass, but I don't want to extend the ComputePass for a general usage. I kept this because I was not clear if we still want to go this way until today's discussion.
It is now safe to revert to ComputePass as we now only dispatch once.

File: Gems/Terrain/Assets/Shaders/Terrain/ClipmapComputeHelpers.azsli
Comment (jiaweig-amzn) [reply Gems/Terrain/Assets/Shaders/Terrain/ClipmapComputeHelpers.azsli:84]: Due to toroidal addressing, the minimum translation to ensure distance to non-negative is (distance + 3 * halfClipmapSize). I'll add a comment here to explain. But we will also change this in the next iteration as well.

File: Gems/Terrain/Assets/Shaders/Terrain/ClipmapComputeHelpers.azsli
Comment (jiaweig-amzn) [reply Gems/Terrain/Assets/Shaders/Terrain/ClipmapComputeHelpers.azsli:86]: Because radius is half of the clipmap size. The number is easier to understand in the configuration. I'll put this into a function and add comments to explain.

**Review by antonmic — APPROVED** [review]

**Pull request merged**
